### PR TITLE
[NEURON][TF][BUILD] Add PT 1.13.0 training with Neuron SDK 2.9.1

### PR DESCRIPTION
**Pull request by catalin-manciu-aws**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description

Adds NeuronX Pytorch 1.13 container based on Neuron SDK 2.9.1 featuring. EC2 and ECS tests were ran on inf2 and trn1. Sagemaker tests are currently disabled until https://github.com/aws/deep-learning-containers/pull/2885 is merged. 

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [x] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: dlc_developer_config.toml [hunk 1]
```diff
@@ -44,7 +47,7 @@ do_build = true
 ### On by default
 sanity_tests = true
   safety_check_test = false
-  ecr_scan_allowlist_feature = false
+  ecr_scan_allowlist_feature = true
```
Comment (shantanutrip) [inline dlc_developer_config.toml:50]: Lets mark this as false

File: pytorch/training/buildspec-1-13-neuronx.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
Comment (shantanutrip) [inline pytorch/training/buildspec-1-13-neuronx.yml:1]: Is this a duplicate of the original `buildspec`?

File: pytorch/training/docker/1.13/py3/sdk2.9.1/Dockerfile.neuronx.os_scan_allowlist.json [hunk 3]
```diff
@@ -0,0 +1,1216 @@
+{
```
Comment (shantanutrip) [inline pytorch/training/docker/1.13/py3/sdk2.9.1/Dockerfile.neuronx.os_scan_allowlist.json:1]: How did we add the allowlist? 

Were the tests failing in case this was not added? Did we see that failure on this PR?

File: test/dlc_tests/conftest.py [hunk 4]
```diff
@@ -58,7 +58,7 @@
     "pytorch_inference_eia",
     "pytorch_inference_neuron",
     "pytorch_inference_neuronx",
-    "pytorch_training_neuron",
+    "pytorch_training_neuronx",
```
Comment (shantanutrip) [inline test/dlc_tests/conftest.py:61]: This change would be backward incompatible. For eg., the `pytorch_training_neuron` images will not be processed by the tests anymore. With the release of new SDK, will the older Neuron images become obsolete? 

If that is the case, its fine to rename the fixture.

Comment (catalin-manciu-aws) [reply test/dlc_tests/conftest.py:61]: Unlikely this will be needed again, training images should use the name 'neuronx'

File: pytorch/training/docker/1.13/py3/sdk2.9.1/Dockerfile.neuronx.os_scan_allowlist.json (hunk 3 above)
Comment (catalin-manciu-aws) [reply pytorch/training/docker/1.13/py3/sdk2.9.1/Dockerfile.neuronx.os_scan_allowlist.json:1]: The allowlist is based on the previous releases' allowlist and includes things we don't control (such as pytorch).

File: pytorch/training/buildspec-1-13-neuronx.yml (hunk 2 above)
Comment (catalin-manciu-aws) [reply pytorch/training/buildspec-1-13-neuronx.yml:1]: It's based on the previous releases' buildspec.

File: test/dlc_tests/conftest.py [hunk 5]
```diff
@@ -129,16 +129,21 @@
     "feature_s3_plugin_present": {NightlyFeatureLabel.AWS_S3_PLUGIN_INSTALLED.value},
 }
 
+
 @pytest.fixture(autouse=True)
-def skip_s3plugin_test(request, pytorch_training):
+def skip_s3plugin_test(request):
```
Comment (shantanutrip) [inline test/dlc_tests/conftest.py:134]: A fix has been merged into the master. We can revert these changes and update this branch with the Master.

File: test/dlc_tests/conftest.py [hunk 6]
```diff
@@ -996,6 +1001,7 @@ def lookup_condition(lookup, image):
     """
     # Extract ecr repo name from the image and check if it exactly matches the lookup (fixture name)
     repo_name = get_ecr_repo_name(image)
+    LOGGER.info(f"[logs-ABCD][Displaying] repo_name: {repo_name} lookup: {lookup} image: {image}")
```
Comment (shantanutrip) [inline test/dlc_tests/conftest.py:1004]: Can you please remove these lines in this file that start with `LOGGER.info(f"[logs-ABCD].....` in the next commit? 

I added these lines for logging purposes and did not remove till the fix was in place.

Since the issue is now fixed, can you please remove it?

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 7]
```diff
@@ -256,7 +281,9 @@ def test_pytorch_gloo_inductor_gpu(
     test_cmd = (
         os.path.join(CONTAINER_TESTS_PREFIX, "pytorch_tests", "testPyTorchGlooMpi") + " gloo 1"
     )  # backend, inductor flags
-    execute_ec2_training_test(ec2_connection, pytorch_training, test_cmd, large_shm=True, timeout=1500)
+    execute_ec2_training_test(
```
Comment (shantanutrip) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:284]: Was this automatically done by the formatter?

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 8]
```diff
@@ -58,22 +58,45 @@
 PT_EC2_NEURON_TRN1_INSTANCE_TYPE = get_ec2_instance_type(
     default="trn1.2xlarge", processor="neuron", job_type="training"
 )
+PT_EC2_NEURON_INF2_INSTANCE_TYPE = get_ec2_instance_type(
+    default="inf2.xlarge", processor="neuron", job_type="training"
```
Comment (shantanutrip) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:62]: Why do we call this process as `Neuron` and not `Neuronx`?

File: test/dlc_tests/conftest.py (hunk 5 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/conftest.py]: Done

File: test/dlc_tests/conftest.py (hunk 6 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/conftest.py]: Done

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py (hunk 8 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py]: Done, though i don't think it makes any functional difference

**Review by aws-ceenugal — APPROVED** [review]

**Review by shantanutrip — APPROVED** [review]

**Comment by shantanutrip** [comment]
Can we add a detailed description to the PR and also the tests we ran? Can we also provide a list of tests that ran and the commits that they were run on?
